Added loggamma to known_functions_mpmath

Soniyanayak51 opened the pull request:
Added loggamma to known_functions_mpmath

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #17411 

#### Brief description of what is fixed or changed
Added loggamma to known_functions_mpmath in pycode function to make lambdify loggamma work for mpmath.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
  * Lambdifying an expression with loggamma now works for mpmath.
<!-- END RELEASE NOTES -->

Upabjojr committed:
added support for Sum/Product expansion with exp/log

Upabjojr committed:
minor fixes

Soniyanayak51 committed:
Added loggamma to known_functions_mpmath

Lambdifying an expression now works for mpmath

Soniyanayak51 committed:
Added loggamma to known_functions_mpmath

Lambdifying an expression now works for mpmath

sylee957 commented:
Can you add some test cases in `test_pycode.py`?

Upabjojr committed:
Merge pull request #17412 from Upabjojr/log_product_expand_force

added support for Sum/Product expansion with exp/log

Soniyanayak51 commented:
@sylee957 Alright. On it.

Soniyanayak51 commented:
@sylee sorry I was a little caught up. I am a newcomer and needed some leads in writing the tests. 
I understood I need to write the tests in test_MpmathPrinter() in test_pycode.py
Could you provide me with some insights?

smichr commented:
> Could you provide me with some insights?

I am not that familiar with lambdify, but looking at the test you referenced,
```python
def test_MpmathPrinter():
    p = MpmathPrinter()
    assert p.doprint(sign(x)) == 'mpmath.sign(x)'
    assert p.doprint(Rational(1, 2)) == 'mpmath.mpf(1)/mpmath.mpf(2)'
```
I would guess something like
```python
    assert p.doprint(loggamma(x)) == 'mpmath.loggamma(x)'
```
instead of the current 'mpmath.lgamma(x)'

Soniyanayak51 commented:
@smichr @sylee957 as mentioned in the original issue thread, can we have a function which tests all the functions _known_functions_* dictionaries? It would be nice as many aren't tested for numpy and scipy as well.
If this is not possible, should I add all the tests manually?

Soniyanayak51 committed:
Added tests for _known_functions_mpmath
Added tests in test_MpmathPrinter() for loggamma, fresnels, fresnelc in
test_pycode.py

Soniyanayak51 committed:
Merge branch 'lambdify_loggamma_for_mpmath' of https://github.com/Soniyanayak51/sympy into lambdify_loggamma_for_mpmath

sylee957 commented:
`MpmathPrinter()._kf` or `NumPyPrinter()._kf` may contain the look up table about the functions it can print, but I don't think that it would be easy to find original sympy function to test out from the dictionary keys since they are pure strings, and you may have to consider other complicated things like how many arguments the original sympy functions can take.

So, if you make a more complicated mapping to control all such stuff, I think that it would be as long as repeating the `assert` line or such.
I'm fine with making some boilerplates with the test cases.
I was even not very fond of using the key-value mapping for printing out functions or constants as in many other printers, because we can't see how really the code are covered.

Soniyanayak51 commented:
Allright then. I will complete the tests.

jmig5776 commented on sympy/functions/elementary/exponential.py:
```diff
@@ -768,7 +768,7 @@ def _eval_expand_log(self, deep=True, **hints):
                     return unpolarify(e) * a
         elif isinstance(arg, Product):
-            if arg.function.is_positive:
+            if force or arg.function.is_positive:
```
Can you check for some examples  where `arg.function.is_positive` is `None`. Because Sympy has tri classification i.e. `True, False and None`.
Might be there cases when `args.function.is_positive = None` and it may enter the if and might give the wrong answer.
If you could find some cases then more precisely we can write
```
arg.function.is_positive is True:
```
I had not delved deeper, let me know if you could find such cases or may be I am missing something.

jmig5776 commented:
Something is wrong here why your pull request includes changes of the #17412 by @Upabjojr .

sylee957 commented:
@jmig5776 
I don't think the changes are related here, but from #17412.
Github may still show all the unnecessary diff from the merge. So you may have to rebase instead of merge the master to drop all the diff in #17412.

jmig5776 commented:
> So you may have to rebase instead of merge the master to drop all the diff in #17412.

@Soniyanayak51

Soniyanayak51 commented:
@jmig5776 I noticed and was wondering the same

Soniyanayak51 committed:
Added tests for _known_functions_numpy and _known_functions_scipy_special
in test_pycode.py

Soniyanayak51 commented:
@sylee957 tests are added.

smichr commented:
Something like this can be done to check which names are not the same in math and numpy or mpmath:
```python
>>> from sympy.printing.pycode import _known_functions_math as kf
>>> for k in kf:
...     try:
...         exec(S('p.doprint(%s(x))'%k,{'p':MpmathPrinter()}))
...     except (AttributeError, NameError):
...         print('undefined:',k)
...     except TypeError: pass
('undefined:', 'log2')
('undefined:', 'loggamma')
('undefined:', 'log1p')
>>> for k in kf:
...     try:
...         exec(S('p.doprint(%s(x))'%k,{'p':NumPyPrinter()}))
...     except (AttributeError, NameError):
...         print('undefined:',k)
...     except TypeError: pass
('undefined:', 'ceiling')
('undefined:', 'Sqrt')
('undefined:', 'cosh')
('undefined:', 'hypot')
('undefined:', 'acosh')
('undefined:', 'tan')
('undefined:', 'asin')
('undefined:', 'log')
('undefined:', 'floor')
('undefined:', 'ln')
('undefined:', 'atanh')
('undefined:', 'log10')
('undefined:', 'sin')
('undefined:', 'asinh')
('undefined:', 'log2')
('undefined:', 'atan')
('undefined:', 'factorial')
('undefined:', 'expm1')
('undefined:', 'sinh')
('undefined:', 'cos')
('undefined:', 'tanh')
('undefined:', 'loggamma')
('undefined:', 'atan2')
('undefined:', 'erf')
('undefined:', 'erfc')
('undefined:', 'exp')
('undefined:', 'acos')
('undefined:', 'log1p')
('undefined:', 'gamma')
```

Soniyanayak51 commented:
From travis logs errors are::
assert p.doprint(factorial(x)) == 'scipy.special.factorial(x, exact=true)'
AssertionError

and

assert p.doprint(asinh(x)) == 'numpy.arcshinh(x)'
AssertionError
 
The first one is fixed by removing exact=true from the argument list but all inverse and inverse hyperbolic functions are giving assertion error.
>>> from mpmath import *
>>> print(asinh(x))
2.31243834127275
>>> np.arcsinh(x)
2.3124383412727525

when x = 5 
So should I take upto some number of decimal points for the tests?

sylee957 commented on sympy/printing/tests/test_pycode.py:
```diff
@@ -64,10 +67,20 @@ def test_MpmathPrinter():
+    assert p.doprint(acos(x)) == 'numpy.arccos(x)'
+    assert p.doprint(acosh(x)) == 'numpy.arccosh(x)'
+    assert p.doprint(asin(x)) == 'numpy.arcsin(x)'
+    assert p.doprint(asinh(x)) == 'numpy.arcshinh(x)'
```
```suggestion
    assert p.doprint(asinh(x)) == 'numpy.arcsinh(x)'
```

sylee957 commented on sympy/printing/tests/test_pycode.py:
```diff
@@ -92,6 +105,26 @@ def test_SciPyPrinter():
+    assert p.doprint(chebyshevu(x, y)) == 'scipy.special.eval_chebyu(x, y)'
+    assert p.doprint(legendre(x, y)) == 'scipy.special.eval_legendre(x, y)'
+    assert p.doprint(hermite(x, y)) == 'scipy.special.eval_hermite(x, y)'
+    assert p.doprint(laguerre(x, 0, y)) == 'scipy.special.eval_laguerre(x, y)'
```
```suggestion
    assert p.doprint(laguerre(x, y)) == 'scipy.special.eval_laguerre(x, y)'
```

sylee957 commented on sympy/printing/tests/test_pycode.py:
```diff
@@ -92,6 +105,26 @@ def test_SciPyPrinter():
+    assert p.doprint(loggamma(x)) == 'scipy.special.gammaln(x)'
+    assert p.doprint(digamma(x)) == 'scipy.special.psi(x)'
+    assert p.doprint(RisingFactorial(x, y)) == 'scipy.special.poch(x, y)'
+    assert p.doprint(jacobi(1, x, y, z)) == 'scipy.special.eval_jacobi(1, x, y, z)'
```
```suggestion
    assert p.doprint(jacobi(1, x, y, z, evaluate=False)) == 'scipy.special.eval_jacobi(1, x, y, z)'
```
This function automatically evaluates, so the only way to disable is like this way

sylee957 commented on sympy/printing/tests/test_pycode.py:
```diff
@@ -92,6 +105,26 @@ def test_SciPyPrinter():
+    assert p.doprint(factorial(x)) == 'scipy.special.factorial(x, exact=true)'
+    assert p.doprint(gamma(x)) == 'scipy.special.gamma(x)'
+    assert p.doprint(loggamma(x)) == 'scipy.special.gammaln(x)'
+    assert p.doprint(digamma(x)) == 'scipy.special.psi(x)'
```
I don't think that the any symbolic object for `digamma` exists for SymPy, but a redirective to `polygamma`
It would be more tricky to fix, but it would either be adding `_print_polygamma` to the scipy printer and dispatching the trivial case with `psi`, like how we treat `sqrt` vs `pow`
Or making a change to the sympy `digamma` function itself, to be a full symbolic node which automatically evaluates to `polygamma` if `evaluate=True`, but remain `digamma` if `evaluate=False`.

Soniyanayak51 replied:
Yes I see there's a problem here, thank you for pointing this out.

Soniyanayak51 replied:
@sylee957 should I make changes in the digamma function to evaluate digamma if evaluate=False? Adding _print_polygamma seems a little complicated to me. If I take reference from sqrt and pow, many functions need to be added for digamma as well?

Soniyanayak51 replied:
I would need help with  _print_polygamma or making changes to the existing function.

sylee957 replied:
I have the diff like below, which can resolve the issue related to digamma without adding `_print_polygamma`, but one limitation is that things will not be printed in `digamma` unless the automatic evaluation is manually disabled.

```
$ git diff
diff --git a/sympy/core/tests/test_args.py b/sympy/core/tests/test_args.py
index 38df142210..0d7b2ad263 100644
--- a/sympy/core/tests/test_args.py
+++ b/sympy/core/tests/test_args.py
@@ -2278,6 +2278,11 @@ def test_sympy__functions__special__gamma_functions__polygamma():
     assert _test_args(polygamma(x, 2))


+def test_sympy__functions__special__gamma_functions__digamma():
+    from sympy.functions.special.gamma_functions import digamma
+    assert _test_args(digamma(x))
+
+
 def test_sympy__functions__special__gamma_functions__uppergamma():
     from sympy.functions.special.gamma_functions import uppergamma
     assert _test_args(uppergamma(x, 2))
diff --git a/sympy/functions/special/gamma_functions.py b/sympy/functions/special/gamma_functions.py
index d9bb918e32..d596ad0a90 100644
--- a/sympy/functions/special/gamma_functions.py
+++ b/sympy/functions/special/gamma_functions.py
@@ -965,7 +965,7 @@ def _sage_(self):
         return sage.log_gamma(self.args[0]._sage_())


-def digamma(x):
+class digamma(Function):
     r"""
     The digamma function is the first derivative of the loggamma function i.e,

@@ -993,7 +993,10 @@ def digamma(x):
     .. [2] http://mathworld.wolfram.com/DigammaFunction.html
     .. [3] http://functions.wolfram.com/GammaBetaErf/PolyGamma2/
     """
-    return polygamma(0, x)
+    def __new__(cls, x, evaluate=True):
+        if evaluate:
+            return polygamma(0, x, evaluate=evaluate)
+        return super(digamma, cls).__new__(cls, x, evaluate=False)


 def trigamma(x):
diff --git a/sympy/printing/pycode.py b/sympy/printing/pycode.py
index 2d908b4236..744132076a 100644
--- a/sympy/printing/pycode.py
+++ b/sympy/printing/pycode.py
@@ -818,6 +818,7 @@ def _print_CodegenArrayElementwiseAdd(self, expr):
     'factorial': 'factorial',
     'gamma': 'gamma',
     'loggamma': 'gammaln',
+    'polygamma': 'polygamma',
     'digamma': 'psi',
     'RisingFactorial': 'poch',
     'jacobi': 'eval_jacobi',
diff --git a/sympy/printing/tests/test_pycode.py b/sympy/printing/tests/test_pycode.py
index 30af61da9e..cb54fe751a 100644
--- a/sympy/printing/tests/test_pycode.py
+++ b/sympy/printing/tests/test_pycode.py
@@ -114,6 +114,9 @@ def test_SciPyPrinter():
     assert p.doprint(besselk(x, y)) == 'scipy.special.kv(x, y)'
     assert p.doprint(factorial(x)) == 'scipy.special.factorial(x)'
     assert p.doprint(gamma(x)) == 'scipy.special.gamma(x)'
+    assert p.doprint(polygamma(x, y)) == 'scipy.special.polygamma(x, y)'
+    assert p.doprint(digamma(x)) == 'scipy.special.polygamma(0, x)'
+    assert p.doprint(digamma(x, evaluate=False)) == 'scipy.special.psi(x)'
     assert p.doprint(loggamma(x)) == 'scipy.special.gammaln(x)'
     assert p.doprint(RisingFactorial(x, y)) == 'scipy.special.poch(x, y)'
     assert p.doprint(jacobi(1, x, y, z, evaluate=False)) == 'scipy.special.eval_jacobi(1, x, y, z)'

```

sylee957 commented on sympy/printing/tests/test_pycode.py:
```diff
@@ -92,6 +105,26 @@ def test_SciPyPrinter():
+    assert p.doprint(bessely(x, y)) == 'scipy.special.yv(x, y)'
+    assert p.doprint(besseli(x, y)) == 'scipy.special.iv(x, y)'
+    assert p.doprint(besselk(x, y)) == 'scipy.special.kv(x, y)'
+    assert p.doprint(factorial(x)) == 'scipy.special.factorial(x, exact=true)'
```
```suggestion
    assert p.doprint(factorial(x)) == 'scipy.special.factorial(x)'
```

Soniyanayak51 committed:
Added changes and removed errors from tests.
To add: test for digamma.

Soniyanayak51 committed:
Changed digamma to a symbolic node in gamma_functions.py
which automatically evaluates to polygamma if evaluate=True, but remain
digamma if evaluate=False. Added tests for digamma and polygamma in
test_pycode.py.

oscargus commented on sympy/functions/elementary/exponential.py:
```diff
@@ -108,12 +108,12 @@ def _eval_power(self, other):
         return Pow._eval_power(Pow(b, e, evaluate=False), other)
 
     def _eval_expand_power_exp(self, **hints):
+        from sympy import Sum, Product
```
I think that everything in this file comes from another PR that has been merged into this PR. But I agree with you that it should have been imported at the top.

oscargus commented on sympy/functions/special/gamma_functions.py:
```diff
@@ -993,7 +993,10 @@ def digamma(x):
-    return polygamma(0, x)
+    def __new__(cls, x, evaluate=True):
+        if evaluate:
+            return polygamma(0, x, evaluate=evaluate)
```
Is it possible to evaluate digamma into polygamma later?

Soniyanayak51 replied:
@oscargus I didn't understand. The idea was to keep it polygamma when evaluate=True and digamma when evaluate=False.

sylee957 replied:
Calling `digamma(x, evaluate=False).doit()` will restore to `polygamma` since it reconstructs the node.
My idea here is that it is better to have some options to bypass automatic evaluation.

asmeurer commented:
I'm not sure why there's a separate list in the printer, but should this instead be added to MPMATH_TRANSLATIONS in lambdify.py? Note that that list is also used by the evalf mapper in Function.

asmeurer commented:
Oh is it the difference between mpmath functions that are different and mpmath functions that have the same name? Is it really necessary to have every mpmath function listed in that dictionary in the printer for it to work?

oscargus commented:
> Is it really necessary to have every mpmath function listed in that dictionary in the printer for it to work?

When I read the code, it seemed like there is some attempt to import all functions and match. But it doesn't seems to work that well/at all.

Also, it is not clear (to me) how lambify vs. code printers actually work. See this test that I added in #17394

```
@XFAIL
def test_beta_math():
    # Not clear why it is not working since pycode(beta(x, y))
    # gives 'math.gamma(x)*math.gamma(y)/math.gamma(x + y)'

    f = beta(x, y)
    F = lambdify((x, y), f, modules='math')

    assert abs(beta(1.3, 2.3) - F(1.3, 2.3)) <= 1e-10
```
lambdify doesn't seem to call the code printer (in a way that I would expect...).

So not really an answer, more a confirmation that it is not really obvious...

sylee957 commented:
@oscargus 
I think that the problem is because `lambdify` uses printers with `allow_unknown_functions=True`, and I've also checked that it prints in wrong way for `math` printer if the setting is enabled.
I have a fix in #17501

czgdp1807 commented:
@Soniyanayak51 Any news on this? Please let me know if someone take this over. Thanks for your contributions.

oscargus commented:
Closing this as it was fixed in the merged  #19913
